zetaclient/chains/solana/observer/inbound.go

```diff
@@ -281,16 +288,39 @@ func FilterInboundEvents(
 				logger.Info().
 					Str(logs.FieldMethod, "FilterInboundEvents").
 					Str("signature", tx.Signatures[0].String()).
-					Int("instruction", i).
+					Str("location", location).
 					Msg("call detected")
+				return nil
 			}
 		} else {
 			logger.Warn().
 				Str(logs.FieldMethod, "FilterInboundEvents").
 				Str("signature", tx.Signatures[0].String()).
-				Int("instruction", i).
+				Str("location", location).
 				Msg("multiple calls detected")
 		}
+
+		return nil
+	}
+
+	// parse top-level instructions
+	for i, instruction := range tx.Message.Instructions {
+		err := parseInstruction(instruction, fmt.Sprintf("top-level instruction %d", i))
+		if err != nil {
+			return nil, err
+		}
+	}
```

Comment:
if an outer instruction is found , do we need to scan for inner after that ? 

I guess only one instruction can be processed per TX , cause the second one would fail at `IsFinalizedInbound` check. Not 100% sure , maybe we can test this ?

Reply:
Yeah. The original idea was to keep the logic consistent with EVM chains, one gas/erc20 deposit per tx, so here is one gas/spl deposit tx. The cap on num of events was to prevent spamming. I think keep one for each event type is okay.

Reply:
The event index is set to 0 irrespective of the type, so even if we vote on multiple events, only the first one would be processed, right?

The key for IsFinalized being `chainID, inboundHash, eventIndex` 
So we would have multiple msg.Digest , but same `IsFinalizedKey`

Reply:
That's a good point. We probably have to use actual instruction index inside the transaction as the event index.